docs/weave-tutorial/06-buckets.md

```diff
@@ -0,0 +1,131 @@
+---
+id: buckets
+title: Buckets
+sidebar_label: Buckets
+---
+
+> [PR#8](https://github.com/iov-one/tutorial/pull/8): _Create buckets_
+\
+[PR#9](https://github.com/iov-one/tutorial/pull/9): _Add indexer for market using marketID, askTicker, bidTicker_
+
+When running your handlers, you get access to the root **KVStore**, which is an abstraction level similar to boltdb or leveldb. It is our data warehouse. An extension can opt-in to using one or more Buckets to store the data. Buckets offer the following advantages:
+
+- Isolation between extensions (each Bucket has a unique prefix that is transparently prepended to the keys)
+- Type safety (enforce all data stored in a Bucket is the same type, to avoid parse errors later on)
+- Indexes (Buckets are well integrated with the secondary indexes and keep them in sync every time data is modified)
+- Querying (Buckets can easily register query handlers including prefix queries and secondary index queries)
+
+All extensions from weave use Buckets, so for compatibility as well as the features, **please use Buckets in your app**, unless you have a very good reason not to (and know what you are doing).
+
+To do so, you will have to wrap your state data structures into [Objects](https://godoc.org/github.com/iov-one/weave/orm#Object). The simplest way is to use `SimpleObj`:
```

Comment:
We have worked on improving bucket implementation for quite some time and never fully finished. There are several issues with discussion and design ideas. ([1](https://github.com/iov-one/weave/issues/824), [2](https://github.com/iov-one/tutorial/issues/16), [3](https://github.com/iov-one/weave/issues/894)). This process was never finished. As a result, we have code that is using more than 3 bucket implementations. The main ones I believe are
- The oldest [`orm.Bucket`](https://github.com/iov-one/weave/blob/master/orm/bucket.go#L33-L47) that you are describing here. There are also minor derivation of this interface.
- [`orm.ModelBucket`](https://github.com/iov-one/weave/blob/master/orm/model_bucket.go#L33-L72) that we started to migrate to. I believe it is an improvement over `orm.Bucket`.
- [`morm.ModelBucket`](https://github.com/iov-one/tutorial/blob/master/morm/model_bucket.go#L40-L91) that seems to be an extended `orm.ModelBucket`

I think that the first bucket implementation `orm.Bucket` should not be recommended. It requires more code and work to use it. I would encourage the use of `orm.ModelBucket` as it is a small interface that covers all important use cases.